lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+		}
+		bav._setBalanceEntryMappings(operationData.PrevTransactorBalanceEntry, true)
+	}
+
+	// Reverting the DeSo addition should not result in more coins.
+	if profilePKID.IsZeroPKID() {
+		// Revert the DeSo add.
+		operationData = utxoOpsForTxn[operationIndex]
+		if operationData.Type != OperationTypeAddBalance {
+			return fmt.Errorf("_disconnectCoinLockup: Trying to revert OperationTypeAddBalance "+
+				"but found type %v", operationData.Type)
+		}
+		if !bytes.Equal(operationData.BalancePublicKey, currentTxn.PublicKey) {
+			return fmt.Errorf("_disconnectCoinLockup: Trying to revert OperationTypeAddBalance " +
+				"but found mismatched public keys")
+		}
+		err := bav._unAddBalance(operationData.BalanceAmountNanos, operationData.BalancePublicKey)
+		if err != nil {
+			return errors.Wrapf(err, "_disconnectCoinLockup: Problem unAdding balance of %v for the "+
+				"transactor", operationData.BalanceAmountNanos)
+		}
+		operationIndex--
+		if operationIndex < 0 {
+			return fmt.Errorf("_disconnectCoinLockup: Trying to revert OperationTypeCoinUnlock " +
+				"but malformed utxoOpsForTxn")
+		}
+	}
+
+	// By here we only need to disconnect the basic transfer associated with the transaction.
+	basicTransferOps := utxoOpsForTxn[:operationIndex]
+	err := bav._disconnectBasicTransfer(currentTxn, txnHash, basicTransferOps, blockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectCoinLockup")
+	}
+	return nil
+}
+
+//
+// DB FLUSHES
+//
+
+func (bav *UtxoView) _flushLockedBalanceEntriesToDbWithTxn(txn *badger.Txn, blockHeight uint64) error {
+	// Go through all entries in the LockedBalanceEntryMapKeyToLockedBalanceEntry map.
+	for lockedBalanceEntryMapKeyIter, lockedBalanceEntry := range bav.LockedBalanceEntryKeyToLockedBalanceEntry {
+		lockedBalanceEntryKey := lockedBalanceEntryMapKeyIter
+
+		// Sanity check the key computed from the lockedBalanceEntry is equal
+		// to the lockedBalanceEntryKey that maps to that entry.
+		lockedBalanceEntryKeyInEntry := lockedBalanceEntry.ToMapKey()
+		if lockedBalanceEntryKeyInEntry != lockedBalanceEntryKey {
+			return fmt.Errorf("_flushLockedBalanceEntriesToDbWithTxn: LockedBalanceEntry has "+
+				"LockedBalanceEntryKey: %v, which doesn't match the LockedBalanceEntryMapKeyToLockedBalanceEntry map key %v",
+				&lockedBalanceEntryKeyInEntry, &lockedBalanceEntry)
+		}
+
+		// Delete the existing mappings in the db for this LockedBalanceEntry.
+		// They will be re-added if the corresponding entry in memory has isDeleted=false.
+		if err := DbDeleteLockedBalanceEntryWithTxn(txn, bav.Snapshot, *lockedBalanceEntry); err != nil {
+			return errors.Wrapf(
+				err, "_flushLockedBalanceEntriesToDbWithTxn: Problem deleting mappings "+
+					"for LockedBalanceEntry: %v", &lockedBalanceEntryKey)
+		}
+	}
+	for _, lockedBalanceEntry := range bav.LockedBalanceEntryKeyToLockedBalanceEntry {
+		if lockedBalanceEntry.isDeleted || lockedBalanceEntry.BalanceBaseUnits.IsZero() {
```

Comment:
I see you're skipping based on IsZero() here. I think this will work, but we tend to prefer not doing implicit deletions like this because it can lead to unexpected and hard-to-track bugs. Instead, I would try and make it so that your code marks things with zero balance with isDeleted=true before it gets here.